### ISSUE TITLE
[Qwen3 MoE] Make auxiliary-loss-free expert-bias updates explicit and migration-safe

### ISSUE THREAD

**Issue by yuweih205** [pytorch/torchtitan#4394]
## Summary

This issue is **not** asking to revert #3666. That PR fixed a real omission: Qwen3 had a default `load_balance_coeff=1e-3`, but the `register_moe_load_balancing_hook` was not registered, so the expert bias was not actually updated during training.

The remaining concern is that the behavior change is implicit. The implementation does not clearly separate:

1. whether a model owns, loads, and uses the persistent `expert_bias_E`; and
2. whether training updates `expert_bias_E` from router token statistics.

Those are different compatibility choices. A user may want to load a checkpoint containing a non-zero expert bias while freezing it, or may need to reproduce a pre-#3666 run exactly.

This is intentionally an issue/RFC rather than a revert or an immediate API PR: the desired compatibility contract and default behavior should be agreed on before changing the model registry.

## Reproducibility impact

The following controlled comparison used the same initial checkpoint, data order, `seed=42`, deterministic execution, and the same model/runtime settings:

- Qwen3-235B-A22B
- Torch 2.14, FA3, unfused QKV, ordinary cross-entropy, standard EP
- old TorchTitan commit `9f055b433d326dcb849d6cb29ab7c104ef315df8`
- current TorchTitan `main` commit `fbceec07e35cc2dd69c55df7d580f19381421a12`
- both runs completed 1,000 steps

The old/current comparison showed:

| Metric | Result |
| --- | --- |
| First divergent loss step | 2 |
| First divergent `grad_norm` step | 19 |
| Loss MAE | `0.04191409` |
| Loss RMSE | `0.05476717` |
| Maximum absolute loss difference | `0.12922001` at step 955 |
| Step-1000 absolute loss difference | `0.08585262` |
| Step-1000 loss | `7.260471` (`main`) vs. `7.174619` (`old`) |

The same step-2 loss divergence was reproduced with Qwen3-30B-A3B.

## Ablation evidence

The experiments provide evidence that #3666 and #3386 contribute different effects: #3666 changes the training state used by the next router invocation, while #3386 changes the distributed floating-point execution path.

| Comparison | Loss result | `grad_norm` result |
| --- | --- | --- |
| #3666 enabled vs. disabled | First divergence at step 2; max absolute difference `0.10892582` at step 955 | First divergence at step 17; max absolute difference `5.3125` at step 720 |
| Current `main` vs. old, #3666 update disabled on `main` | MAE `0.01192084`; RMSE `0.01536384`; max `0.03840256` at step 513; final difference `0.01884270` | Reduced relative to the enabled-hook comparison |
| #3386 boundary comparison, with #3666 disabled on both sides | First divergence at step 2; max `1.43051e-4` at step 42 | First divergence at step 17; max `0.125` |
| #3386 and #3666 both disabled vs. old | `1000/1000` bitwise exact; MAE/RMSE/max all zero | `1000/1000` bitwise exact |

In the combined ablation, batch fingerprints also matched for all 1,000 steps. This is a causal boundary for the old/current comparison, not a claim that #3666 explains every numerical difference observed in other configurations.

### Scope boundary: composite dense-path experiments

A separate experiment measured a much larger drift from the **combination** `FlexAttention + fused QKV + ChunkedLoss` versus `SDPA + unfused QKV + ordinary CE`: loss MAE `0.219222` and maximum absolute difference `0.665282` over 200 paired steps. These choices correspond to the modern paths associated with #3571, #2878, and #2937, but this number must not be attributed to #3666 or to any single PR.

The available feature-level measurements were:

| Matrix configuration | Paired window | Loss MAE | Max absolute loss difference |
| --- | ---: | ---: | ---: |
| `FA3 + fused QKV + CE` | through step 279 | `0.0001633` | `0.0008516` |
| `FA3 + unfused QKV + ChunkedLoss` | through step 279 | `0.0002068` | `0.0010004` |
| `FA3 + fused QKV + ChunkedLoss` | through step 279 | `0.0001707` | `0.0008869` |
| old FlexAttention vs. old SDPA | 1,000 steps | `0.00222` | `0.03046` |

These are per-path diagnostics from different comparison windows, not three strict one-at-a-time removals or a complete `2^3` factorial experiment. The corresponding matrix runs were not all successful to the same final step; the values above are paired through step 279. In the aligned dense-model comparison, replacing the three modern choices with `SDPA + unfused QKV + ordinary CE` restored bitwise overlap. That limited statement should not be generalized to the Qwen3 MoE result: the 1,000-step MoE bitwise closure required disabling #3386 and #3666 while already using `FA3 + unfused QKV + plain CE`.

## Current behavior

The relevant implementation is split across:

- `torchtitan/models/qwen3/__init__.py`: the Qwen3 registry unconditionally sets `post_optimizer_build_fn=register_moe_load_balancing_hook`;
- `torchtitan/models/common/moe.py`: `load_balance_coeff` controls whether the persistent `expert_bias_E` buffer is created;
- `torchtitan/components/optimizer/optimizer.py`: the optimizer pre-hook updates the bias and clears the accumulated token counts.

The current Qwen3 registry uses:

```python
post_optimizer_build_fn=register_moe_load_balancing_hook
```

and the common MoE configuration defaults to:

```python
load_balance_coeff: float | None = 1e-3
```

`expert_bias_E` is a persistent buffer passed into routing. The hook updates it before each optimizer step from the accumulated `tokens_per_expert_E`. Thus, loading the same checkpoint and consuming the same next batch can produce a different next-step route depending on whether the hook is registered.

The Qwen3 README currently describes the auxiliary-loss-free load-balancing algorithm but does not specify:

- the default coefficient and whether `None` disables updates;
- the optimizer pre-hook timing;
- how token statistics behave with gradient accumulation;
- whether checkpoint metadata records the update mode;
- how to resume or reproduce a run created before #3666.

## Requested behavior

Could maintainers advise on, or help implement, the following contract?

1. **Separate ownership/loading from mutation.** An explicit `update_expert_bias` (or equivalent) control should determine whether the bias is updated, without disabling the parameter/buffer or preventing checkpoint loading.
2. **Define freeze semantics.** With updates disabled, a non-zero `expert_bias_E` loaded from a checkpoint should remain in use and should not be silently reset.
3. **Document the update boundary.** Specify whether the update happens once per optimizer step, which token statistics it consumes under gradient accumulation, and its ordering relative to the optimizer update.
4. **Make migration observable.** Document the #3666 training-semantics change and, if practical, record the effective update mode/coefficient in saved configuration or checkpoint metadata.
5. **Add regression tests.**
   - non-zero `expert_bias_E` checkpoint round-trip;
   - checkpoint/resume with updates enabled;
   - checkpoint/resume with updates disabled/frozen;
   - gradient accumulation and optimizer-step boundary;
   - Qwen3 registry construction with updates enabled and disabled.

A low-risk first step would be a Qwen3 README/release-note clarification plus focused checkpoint and registry tests. The API/config change can follow once the desired compatibility contract is agreed.

## Related work

- #3666 — registers the Qwen3 auxiliary-loss-free load-balancing hook
- #3386 — independently changes MoE sharding and reduction paths and is relevant when interpreting numerical comparisons
- #4393 — adds paired loss/`grad_norm` statistics to make this class of divergence easier to diagnose

**Comment by tianyu-l** [comment]
IIUC Qwen3 only uses aux loss, not aux-loss-free lb coefficient?

The former is dependent on https://github.com/pytorch/torchtitan/pull/3864 and followup from @pianpwk 

The latter can be removed from Qwen 3 after that.

**Comment by yuweih205** [comment]
Okay, thanks! Just to confirm, does this mean the auxiliary-loss-free expert-bias update mechanism will ultimately not be retained for Qwen3?

I ask because I noticed that the [Qwen3 README](https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/qwen3/README.md#L12) currently states that the MoE models use auxiliary-loss-free load balancing.

**Comment by tianyu-l** [comment]
I think our goal is to faithfully follow what the model was trained with. But users should feel free to create recipes to alter configs

**Comment by yuweih205** [comment]
Understood, thanks. Following what the model was trained with is the right default, and with the aux-loss-free coefficient going away for Qwen3 after #3864 the ambiguity this issue was about disappears. Closing.

**Comment by yuweih205** [comment]
Reopening with one new data point that did not exist when I closed this.

#4469 (merged 2026-09-04) updated `torchtitan/models/qwen3/README.md` to say that Qwen3 MoE load balancing "follows the original Qwen3 training recipe (auxiliary loss)". On current `main`, however, the Qwen3 MoE config still inherits the default `load_balance_coeff=1e-3`, so every MoE layer allocates `expert_bias_E`, and `register_moe_load_balancing_hook` is still attached as `post_optimizer_build_fn` in `torchtitan/models/qwen3/__init__.py`. Training therefore still runs the auxiliary-loss-free expert-bias update from #3666, and the README and the code now disagree.

#3864 is still open and does not touch `qwen3`, so the removal @tianyu-l mentioned above has not happened yet. I am happy to send that PR once #3864 lands: drop the expert-bias update from Qwen3 and wire the new aux-loss framework in its place. If someone else already owns that follow-up, please say so and I will stay out of the way. In the meantime, if it helps, I can also send a one-line README fix so the docs describe what the code currently does.
